rfcs/remote_channel.md

```diff
@@ -0,0 +1,616 @@
+an error to use the same UUID for multiple contexts.
+
+#### Backend
+
+The message passing backend is based on a Python
+[Queue](https://docs.python.org/3/library/queue.html) object per
+channel, stored in the server stash. The advantage of using queues
+here is that they are designed to allow blocking reads. This means we
+don't have to use a polling API but can, on the server side, wait for
+a message to be added to a channel, and immediately forward the
+message to the client. Because the stash itself runs in a single
+Python process and communicates with the server processes via IPC, it
+is sufficient to use a thread-based `Queue` rather than requiring a
+full process per queue.
+
+#### Low-Level API
+
+The low-level API for messaging is based on the concept of
+Multiple-Producer-Single-Consumer (MPSC) channels. Each channel is
+identified by a UUID. When a channel is created a pair of objects are
+returned, a `RecvChannel` to read from the channel and a `SendChannel`
+to write to the channel. The `SendChannel` may be cloned to create
+new writers, but only a single `RecvChannel` per channel is
+permitted. This limitation is deigned to enforce good design in which
+there are no races between different readers to get a message. In
+particular, the intended use is not for broadcasting tasks to multiple
+consumers, but to message a specific browsing context. Browsing
+contexts will be able to create a `RecvChannel` with the UUID in their
+URL and use that to receive messages from other contexts.
+
+A `SendChannel` has a `send(obj)` method. This causes the object to be
+encoded, first using the remote object serialization (see below) and
+then as a JSON string, before being sent to the queue.
+
+A `RecvChannel` must call its async `connect()` method to start
+receiving messages. Messages are first deserialized from JSON and then
+undergo remote object serialization. Once connected, the
+`RecvChannel` acts like an event target; consumers can call the
+`addEventListener(fn)` method to add a callback function when a
+message is received. Alternatively the async `next()` function will
+return the next message to be received.
+
+The implementation of channels is based on websockets, with one
+websocket per `Channel` object (i.e. a different websocket is used for
+a `SendChannel` vs a `RecvChannel` even when they correspond to the
+same underlying queue). On the backend the websocket handler function
+for a `SendChannel` listens for incoming messages and writes them to
+the corresponding `Queue`, whilst the websocket handler for a
+`RecvChannel` essentially blocks on the `get()` method for the `Queue`
+and writes and message it receives to the socket. In detail some
+additional care is needed to ensure that the websocket handler
+functions shutdown when the socket is closed.
+
+#### Remote Object Serialization
+
+In order to allow passing complex JavaScript objects between contexts,
```

Comment:
nit: It might be better to clarify that this serialization is implemented by pure JavaScript (IIUC `RemoteObject.serialize()` in the draft PR) and thus doesn't depend on WebDriver BiDi impl itself.

Reply:
I've clarified this below, let me know if you think more is required.